File: greencoderefactor-java/src/main/java/uclm/esi/alarcos/greenteam/greencoderefactor_java/transformers/sonar/r1481/UnusedLocalVariablesTransformer.java
package uclm.esi.alarcos.greenteam.greencoderefactor_java.transformers.sonar.r1481;

import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.NodeList;
import com.github.javaparser.ast.body.MethodDeclaration;
import com.github.javaparser.ast.body.VariableDeclarator;
import com.github.javaparser.ast.expr.*;
import com.github.javaparser.ast.stmt.*;
import com.github.javaparser.ast.visitor.ModifierVisitor;
import com.github.javaparser.ast.visitor.Visitable;
import uclm.esi.alarcos.greenteam.greencoderefactor_java.transformers.BaseTransformer;

import java.util.HashSet;
import java.util.Set;

/**
 * Transformer Sonar R1481:
 */
public class UnusedLocalVariablesTransformer extends BaseTransformer {

    @Override
    public CompilationUnit transform(CompilationUnit cu) {

        ModifierVisitor<Void> transformer = new ModifierVisitor<>() {

            @Override
            public Visitable visit(MethodDeclaration md, Void arg) {
                if (md.getBody().isEmpty()) return md;

                boolean changed = true;
                while (changed) {
                    BlockStmt originalBody = md.getBody().orElse(new BlockStmt());
                    Set<String> usedVars = collectUsedVariables(originalBody);
                    BlockStmt newBody = removeUnusedAssignments(originalBody, usedVars);

                    changed = !originalBody.equals(newBody);
                    md.setBody(newBody);
                }

                if (md.getBody().isPresent() && md.getBody().get().isEmpty()) {
                    BlockStmt body = new BlockStmt();
                    body.addStatement(new EmptyStmt());
                    md.setBody(body);
                }

                return super.visit(md, arg);
            }

            private Set<String> collectUsedVariables(BlockStmt body) {
                Set<String> used = new HashSet<>();
                body.walk(NameExpr.class, nameExpr -> used.add(nameExpr.getNameAsString()));
                return used;
            }

            private BlockStmt removeUnusedAssignments(BlockStmt body, Set<String> usedVars) {
                BlockStmt newBody = new BlockStmt();

                for (Statement stmt : body.getStatements()) {
                	
                    if (stmt.isBlockStmt()) {
                        newBody.addStatement(removeUnusedAssignments(stmt.asBlockStmt(), usedVars));
                        continue;
                    }

                    if (stmt.isIfStmt()) {
                        IfStmt ifStmt = stmt.asIfStmt();
                        ifStmt.setThenStmt(removeUnusedAssignments(getBlock(ifStmt.getThenStmt()), usedVars));
                        ifStmt.getElseStmt().ifPresent(elseStmt ->
                                ifStmt.setElseStmt(removeUnusedAssignments(getBlock(elseStmt), usedVars))
                        );
                        newBody.addStatement(ifStmt);
                        continue;
                    }

                    if (stmt.isExpressionStmt()) {
                        Expression expr = stmt.asExpressionStmt().getExpression();

                        if (expr.isVariableDeclarationExpr()) {
                            VariableDeclarationExpr decl = expr.asVariableDeclarationExpr();
                            NodeList<VariableDeclarator> newVars = new NodeList<>();

                            for (VariableDeclarator var : decl.getVariables()) {
                                String varName = var.getNameAsString();
                                if (!usedVars.contains(varName)) {
                                    if (var.getInitializer().isPresent()
                                            && var.getInitializer().get().isMethodCallExpr()) {
                                        newBody.addStatement(new ExpressionStmt(var.getInitializer().get()));
                                        logRule(String.format(
                                                "%s: Removed unused variable '%s' but kept call at line %d",
                                                this.getClass().getSimpleName(),
                                                varName,
                                                stmt.getBegin().map(p -> p.line).orElse(-1)
                                        ));
                                    } else {
                                        logRule(String.format(
                                                "%s: Removed unused local variable '%s' at line %d",
                                                this.getClass().getSimpleName(),
                                                varName,
                                                stmt.getBegin().map(p -> p.line).orElse(-1)
                                        ));
                                    }
                                } else {
                                    newVars.add(var);
                                }
                            }

                            if (!newVars.isEmpty()) {
                                VariableDeclarationExpr newDecl = new VariableDeclarationExpr(newVars);
                                newBody.addStatement(new ExpressionStmt(newDecl));
                            }

                            continue;
                        }

                        if (expr.isAssignExpr()) {
                            AssignExpr assign = expr.asAssignExpr();
                            if (assign.getTarget().isNameExpr()) {
                                String varName = assign.getTarget().asNameExpr().getNameAsString();
                                if (!usedVars.contains(varName)) {
                                    logRule(String.format(
                                            "%s: Removed unused assignment to '%s' at line %d",
                                            this.getClass().getSimpleName(),
                                            varName,
                                            stmt.getBegin().map(p -> p.line).orElse(-1)
                                    ));
                                    if (assign.getValue().isMethodCallExpr()) {
                                        newBody.addStatement(new ExpressionStmt(assign.getValue()));
                                    }
                                    continue;
                                }
                            }
                        }

                        if (expr.isAssignExpr() && expr.asAssignExpr().getOperator() != AssignExpr.Operator.ASSIGN) {
                            AssignExpr assign = expr.asAssignExpr();
                            if (assign.getTarget().isNameExpr()) {
                                String varName = assign.getTarget().asNameExpr().getNameAsString();
                                if (!usedVars.contains(varName)) {
                                    logRule(String.format(
                                            "%s: Removed unused augmented assignment '%s' at line %d",
                                            this.getClass().getSimpleName(),
                                            varName,
                                            stmt.getBegin().map(p -> p.line).orElse(-1)
                                    ));
                                    continue;
                                }
                            }
                        }

                        newBody.addStatement(stmt);
                        continue;
                    }

                    newBody.addStatement(stmt);
                }

                if (newBody.isEmpty()) {
                    newBody.addStatement(new EmptyStmt());
                }

                return newBody;
            }

            private BlockStmt getBlock(Statement stmt) {
                if (stmt.isBlockStmt()) {
                    return stmt.asBlockStmt();
                } else {
                    BlockStmt block = new BlockStmt();
                    block.addStatement(stmt);
                    return block;
                }
            }
        };

        transformer.visit(cu, null);
        return cu;
    }
}

File: greencoderefactor-python/trasnsformers/sonar/r1481/unusedlocalvariablestransformer.py
from trasnsformers.basetransformer import BaseTransformer
import ast


class UnusedLocalVariablesTransformer(BaseTransformer):
    """
    Sonar rule R1481.
    """

    def __init__(self):
        super().__init__()

    def visit_FunctionDef(self, node):
        changed = True
        while changed:
            self.used_vars = self._collect_used_names(node.body)
            new_body = self._remove_unused_all(node.body, self.used_vars)
            changed = len(new_body) != len(node.body)
            node.body = new_body
        if not node.body:
            node.body = [ast.Pass()]
        return node

    def _remove_unused_all(self, stmts, used_vars, global_vars=None):
        if global_vars is None:
            global_vars = set()

        new_body = []
        for stmt in stmts:
            if hasattr(stmt, "body") and isinstance(stmt.body, list):
                stmt.body = self._remove_unused_all(stmt.body, used_vars, global_vars)
                if hasattr(stmt, "orelse") and isinstance(stmt.orelse, list):
                    stmt.orelse = self._remove_unused_all(stmt.orelse, used_vars, global_vars)

            if isinstance(stmt, (ast.Global, ast.Nonlocal)):
                global_vars.update(stmt.names)
                new_body.append(stmt)
                continue

            if isinstance(stmt, ast.Assign):
                if len(stmt.targets) == 1 and isinstance(stmt.targets[0], ast.Name):
                    var_name = stmt.targets[0].id
                    if var_name not in used_vars and var_name not in global_vars:
                        value = stmt.value
                        if isinstance(value, ast.Call):
                            new_body.append(ast.Expr(value=value))
                            self.log_rule(
                                f"{self.__class__.__name__}: Removed unused variable '{var_name}' but kept call at line {getattr(stmt, 'lineno', '?')}"
                            )
                        else:
                            self.log_rule(
                                f"{self.__class__.__name__}: Removed unused variable '{var_name}' at line {getattr(stmt, 'lineno', '?')}"
                            )
                        continue

            elif isinstance(stmt, ast.AugAssign):
                if isinstance(stmt.target, ast.Name):
                    var_name = stmt.target.id
                    if var_name not in used_vars and var_name not in global_vars:
                        self.log_rule(
                            f"{self.__class__.__name__}: Removed unused variable operation '{var_name}' at line {getattr(stmt, 'lineno', '?')}"
                        )
                        continue

            elif isinstance(stmt, ast.Delete):
                new_targets = [
                    target for target in stmt.targets
                    if not (isinstance(target, ast.Name) and target.id not in used_vars and target.id not in global_vars)
                ]
                if not new_targets:
                    self.log_rule(
                        f"{self.__class__.__name__}: Removed del of unused variable(s) at line {getattr(stmt, 'lineno', '?')}"
                    )
                    continue
                else:
                    stmt.targets = new_targets

            new_body.append(stmt)

        if not new_body:
            new_body = [ast.Pass()]

        return new_body

    def _collect_used_names(self, stmts):
        used = set()

        class NameCollector(ast.NodeVisitor):
            def visit_Name(self, node):
                if isinstance(node.ctx, ast.Load):
                    used.add(node.id)

        for stmt in stmts:
            NameCollector().visit(stmt)
        return used
